### ISSUE TITLE
Hurdled models don't provide backtransformed estimates for the hu parts of the model when using emtrends(..., type = "response")

### ISSUE THREAD

**Issue by andrewheiss** [paul-buerkner/brms#1349]
With hurdle models, `emmeans::emmeans()` and `emmeans::emtrends()` are able to deal with both the `mu` and the `hu` parts of the model, and `emmeans()` is able to back-transform estimates from logged to unlogged values and from logits to proportions automatically. `emtrends()`, however currently doesn't seem to back-transform the hurdle (`hu`) part of the model.

Here's an annotated reproducible example:

```r
library(tidyverse)
library(gapminder)
library(brms)
library(emmeans)

# Data
set.seed(1234)
gapminder <- gapminder::gapminder |> 
  filter(continent != "Oceania") |> 
  # Make a bunch of GDP values 0 based on life expectancy, so we know the 0 process
  mutate(prob_zero = ifelse(lifeExp < 50, 0.3, 0.02),
         will_be_zero = rbinom(n(), 1, prob = prob_zero),
         gdpPercap = ifelse(will_be_zero, 0, gdpPercap)) |> 
  select(-prob_zero, -will_be_zero)

# Make a simple hurdle model
model_hurdle <- brm(
  bf(gdpPercap ~ lifeExp,
     hu ~ lifeExp),
  data = gapminder,
  family = hurdle_lognormal(),
  chains = 4, iter = 2000, warmup = 1000, seed = 1234
)

# Look at the conditional effects of the mu part of the model. 
conditional_effects(model_hurdle, dpar = "mu")
```

![image](https://user-images.githubusercontent.com/73663/167945457-7be441ea-63bb-44b1-aeca-29df059bae28.png)

```r
# This is the slope of the conditional effects plot there ↑ at each of these
# levels of life expectancy. By default it's on the log scale (and thus the same
# at each life expectancy)
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "mu",
           at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30        0.0785    0.0755    0.0815
#>       40        0.0785    0.0755    0.0815
#>       50        0.0785    0.0755    0.0815
#>       60        0.0785    0.0755    0.0815
#>       70        0.0785    0.0755    0.0815
#>       80        0.0785    0.0755    0.0815
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 

# We can back-transform the conditional effects plot by omitting the dpar argument
conditional_effects(model_hurdle)
```

![image](https://user-images.githubusercontent.com/73663/167945530-136a047c-cfca-45a7-bac1-b522130cbab8.png)

```r
# We can back transform the slopes from emmtrends by using `epred = TRUE`. These
# are the slopes at each value of life expectancy, and they get steeper as life
# exepectancy increases. They're on the original unlogged scale.
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "mu", epred = TRUE,
           at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30          26.9      24.0      29.8
#>       40          73.9      69.2      78.7
#>       50         170.3     162.4     178.8
#>       60         373.1     352.1     393.0
#>       70         814.0     746.8     875.3
#>       80        1779.7    1591.7    1969.5
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 

# In theory, using `type = "response"` should also back transform to the
# non-logged scale(?), but it doesn't work
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "mu", type = "response",
           at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30        0.0785    0.0755    0.0815
#>       40        0.0785    0.0755    0.0815
#>       50        0.0785    0.0755    0.0815
#>       60        0.0785    0.0755    0.0815
#>       70        0.0785    0.0755    0.0815
#>       80        0.0785    0.0755    0.0815
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 
```

So far, everything works for the `mu` part of the model, but back-transforming doesn't happen with the `hu` part:

```r
# Look at the conditional effects of the hurdle (hu) part of the model
conditional_effects(model_hurdle, dpar = "hu")
```

![image](https://user-images.githubusercontent.com/73663/167946414-f0b79a3e-a9cc-418d-871a-bfc69b30e16a.png)

```r
# This is the slope of the conditional effects plot there ↑ at each of these
# levels of life expectancy. By default it's on the logit scale (and thus the
# same at each life expectancy)
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "hu",
           at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30       -0.0989    -0.115   -0.0839
#>       40       -0.0989    -0.115   -0.0839
#>       50       -0.0989    -0.115   -0.0839
#>       60       -0.0989    -0.115   -0.0839
#>       70       -0.0989    -0.115   -0.0839
#>       80       -0.0989    -0.115   -0.0839
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 

# There's no way to back transform these slopes to the probability scale. Adding
# `epred = TRUE` returns predicted slopes for only the non-hurdled part of the
# model (mu), and adding `type = "response"` returns logit values:
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "hu", type = "response",
           at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30       -0.0989    -0.115   -0.0839
#>       40       -0.0989    -0.115   -0.0839
#>       50       -0.0989    -0.115   -0.0839
#>       60       -0.0989    -0.115   -0.0839
#>       70       -0.0989    -0.115   -0.0839
#>       80       -0.0989    -0.115   -0.0839
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 
```

This is *not* the case with emmeans(). Using `type = "response"` *does* return estimated averages as proportions for the hurdled part:

```r
# Logit scale
model_hurdle |> 
  emmeans(~ lifeExp, var = "lifeExp", dpar = "hu",
          at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp emmean lower.HPD upper.HPD
#>       30  0.166    -0.154     0.509
#>       40 -0.820    -1.037    -0.615
#>       50 -1.810    -1.985    -1.648
#>       60 -2.799    -3.041    -2.539
#>       70 -3.787    -4.167    -3.396
#>       80 -4.771    -5.335    -4.263
#> 
#> Point estimate displayed: median 
#> Results are given on the logit (not the response) scale. 
#> HPD interval probability: 0.95 

# Back-transformed results on the proportion scale
model_hurdle |> 
  emmeans(~ lifeExp, var = "lifeExp", dpar = "hu", type = "response",
          at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp response lower.HPD upper.HPD
#>       30   0.5415   0.46168    0.6245
#>       40   0.3057   0.26176    0.3510
#>       50   0.1406   0.12077    0.1613
#>       60   0.0574   0.04418    0.0712
#>       70   0.0222   0.01431    0.0309
#>       80   0.0084   0.00433    0.0130
#> 
#> Point estimate displayed: median 
#> Results are back-transformed from the logit scale 
#> HPD interval probability: 0.95 
```

**Comment by andrewheiss** [comment]
Oh, and @mattansb [suggested](https://twitter.com/mattansb/status/1524402708362539010) I open this issue, so I'm tagging him here :)

**Comment by mattansb** [comment]
A couple of things to note:

1. when `epred = TRUE` the `dpar` (and `nlpar`) arguments are not used. 
So `emtrends(dpar = "mu", epred = TRUE)` is really just `emtrends(epred = TRUE)`
2. In `hurdle_lognormal()`, Mu's link is `link = "identity"`, so `type = "resp"` would be the same as the default.

Finally, ... oops! I said set `type = "resp"` but I meant to set `transform = "resp"`! Sorry!

---

This is what we are trying to re-create with `emtrends()`:

```R
h <- 0.001 * diff(range(gapminder$lifeExp))
grid <- 
  expand.grid(
    Slope = c(0, h),
    lifeExp = seq(30, 80, 10)
  ) |> 
  mutate(
    lifeExp = lifeExp + Slope,
    Slope = NULL
  )
grid
#>    lifeExp
#> 1   30.000
#> 2   30.059
#> 3   40.000
#> 4   40.059
#> 5   50.000
#> 6   50.059
#> 7   60.000
#> 8   60.059
#> 9   70.000
#> 10  70.059
#> 11  80.000
#> 12  80.059

linpred <- posterior_linpred(model_hurdle, newdata = grid, 
                             dpar = "hu", transform = TRUE)

(linpred[,seq(2, 12, by = 2)] - linpred[,seq(1, 12, by = 2)]) |> 
  as.data.frame() |> 
  mutate(across(everything(), ~ .x / h)) |> # rescale to difference quotient
  bayestestR::describe_posterior(test = NULL)
#> Summary of Posterior Distribution
#> 
#> Parameter |    Median |         95% CI
#> --------------------------------------
#> V1        |     -0.02 | [-0.03, -0.02]
#> V2        |     -0.02 | [-0.03, -0.02]
#> V3        |     -0.01 | [-0.01, -0.01]
#> V4        | -5.33e-03 | [-0.01,  0.00]
#> V5        | -2.13e-03 | [ 0.00,  0.00]
#> V6        | -8.12e-04 | [ 0.00,  0.00]

model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "hu", trans = "response",
           at = list(lifeExp = seq(30, 80, 10)))
#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30     -0.024519  -0.02768 -0.021118
#>       40     -0.021026  -0.02585 -0.016437
#>       50     -0.011962  -0.01421 -0.009831
#>       60     -0.005326  -0.00612 -0.004561
#>       70     -0.002126  -0.00262 -0.001635
#>       80     -0.000812  -0.00112 -0.000526
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95
```

**Comment by andrewheiss** [comment]
Awesome, yeah! Ideally something like 

```r
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "hu", transform = "response",
           at = list(lifeExp = seq(30, 80, 10)))
```

would be ideal and would make `emtrends()` behave like `emmeans()`

**Comment by paul-buerkner** [comment]
I am not familiar with `emtrends`. Could you explain to me what would have to be done from the brms side to support this feature (since the `emmeans` method of brms already has all the required arguments).

**Comment by mattansb** [comment]
@andrewheiss Isn't that exactly what I did? 😅
I feel like I'm missing something?

**Comment by andrewheiss** [comment]
I can get the marginal slopes doing it manually with the linpred math, but with emmeans 1.7.4-1 and brms 2.17.0, I'm getting this error when using `emtrends()`:

```r
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "hu", transform = "response",
           at = list(lifeExp = seq(30, 80, 10)))
#> In 'ref_grid()', use 'regrid = ...' rather than 'transform = ...' to avoid this message.
#> In 'ref_grid()', use 'regrid = ...' rather than 'transform = ...' to avoid this message.
#> Error: Cannot coerce 'transform' to a single logical value.
```

I was assuming that error was because of `emtrends()` lack of support for backtransforming hurdle things, but I just discovered that it's because of a change in `ref_grid()` in emmeans 1.7.3. The argument should be regrid now, which *does* work just fine:

```r
model_hurdle |> 
  emtrends(~ lifeExp, var = "lifeExp", dpar = "hu", regrid = "response", 
           at = list(lifeExp = seq(30, 80, 10)))

#>  lifeExp lifeExp.trend lower.HPD upper.HPD
#>       30     -0.024497  -0.02757 -0.020940
#>       40     -0.020910  -0.02576 -0.016613
#>       50     -0.011907  -0.01411 -0.009751
#>       60     -0.005309  -0.00612 -0.004547
#>       70     -0.002129  -0.00262 -0.001638
#>       80     -0.000816  -0.00113 -0.000522
#> 
#> Point estimate displayed: median 
#> HPD interval probability: 0.95 
```

It all works and no changes are needed. Thanks so much @mattansb!
